Adding in a Text Interface for Message PopUp

david-rosales opened the pull request:
I implemented the MobilePrototype of Message PopUp. It looks like a texting interface and allows the user to respond to situations through 'text messaging'. Most of the features from the non Mobile Prototype version are available except for the summary between questions. I felt that the initial point of the summary was to have a pause for the user and show key information. However, the pause already occurred in the Mobile Prototype with users having to press the 'Next Question' button before moving on. Also, while the 'Next Question' button is displayed, the timer stops but is also still displayed. The chatlog also remains so all the key information was present and an extra box with a summary looked very redundant. There are a lot of props being passed into the `MobileInterface` and I decided to leave it like that for now because I want to handle the refactoring in a separate pull request because it will involve similar changes to a refactor in the `ExperiencePage`.

Below is a simple screenshot of one of the changes from the previous prototype. The change is that student info is no longer shown as a dialog when pressing on the face icon for the student because that feature might not seem intuitive. Instead, it shows up as a separate message bubble.
<img width="360" alt="screen shot 2016-07-12 at 1 11 59 pm" src="https://cloud.githubusercontent.com/assets/1490352/16776376/161ff828-4833-11e6-92ed-2ab67a2d64e3.png">

There is now also a feedback and revision option. 
<img width="360" alt="screen shot 2016-07-12 at 1 12 49 pm" src="https://cloud.githubusercontent.com/assets/1490352/16776413/3b38db3e-4833-11e6-9805-022a2ce723fd.png">

Once the initial response is sent, the buttons on the bottom change.
<img width="350" alt="screen shot 2016-07-12 at 1 43 09 pm" src="https://cloud.githubusercontent.com/assets/1490352/16777220/a67cd564-4836-11e6-80a2-11e632bf245a.png">

Also added in the show example button.
<img width="345" alt="screen shot 2016-07-12 at 1 21 57 pm" src="https://cloud.githubusercontent.com/assets/1490352/16777101/3ea530bc-4836-11e6-8947-8dd6581dbd97.png">

Examples are sent as messages until all the examples have been gone through.
<img width="351" alt="screen shot 2016-07-12 at 1 22 13 pm" src="https://cloud.githubusercontent.com/assets/1490352/16777116/53001f40-4836-11e6-978a-5d65922d5711.png">

Once the examples have all been gone through, the 'Show Example' button becomes disabled.
<img width="344" alt="screen shot 2016-07-12 at 1 41 46 pm" src="https://cloud.githubusercontent.com/assets/1490352/16777177/8026d31a-4836-11e6-8942-789e9ea1b9e9.png">

Instead of a toast, an info message is now sent when the user passes the 90 second mark.
<img width="360" alt="screen shot 2016-07-12 at 1 44 30 pm" src="https://cloud.githubusercontent.com/assets/1490352/16777273/dd929f0c-4836-11e6-8768-f14fa1c8588e.png">

Lastly, I decided to change the color of the 'Next Question' button to make it more obvious that the button needs to be pressed to continue. Before, the button would blend into the previous 'Send' button or 'Save Response' button.
<img width="360" alt="screen shot 2016-07-12 at 1 13 16 pm" src="https://cloud.githubusercontent.com/assets/1490352/16776481/844665b2-4833-11e6-8492-801dfe599446.png">


A lot of the props will be simplified in the future along with some of the functions within the components.

david-rosales commented:
@kevinrobinson There's a lot of additions, can you review this when you get a chance?

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,383 @@
+
+
+const styles = {
+  appBar: {
```
Do these all need to be absolutely positioned?  We should try to do some flexbox-ifying :)  I can pair with you on that too if you like.

kevinrobinson replied:
Ping on this question.

kevinrobinson commented on ui/src/message_popup/experience_page.jsx:
```diff
@@ -26,6 +26,8 @@ import * as Api from '../helpers/api.js';
 import FinalSummaryCard from './final_summary_card.jsx';
 import InstructionsCard from './instructions_card.jsx';
 
+import TextingInterface from './mobile_prototype/texting_interface.jsx';
```
It looks like this isn't used anymore?

kevinrobinson commented on ui/src/message_popup/experience_page.jsx:
```diff
@@ -212,11 +214,47 @@ export default React.createClass({
+    const question = questions[questionsAnswered];
+    return ( 
+      <MobileInterface
+        hasStarted = {this.state.hasStarted}
```
As you were saying in person, it'd be good to land the refactoring in https://github.com/mit-teaching-systems-lab/threeflows/pull/67 first so that this can become a simpler interface too.

kevinrobinson commented on ui/src/message_popup/experience_page.jsx:
```diff
@@ -212,11 +214,47 @@ export default React.createClass({
         />
     );
+    
+    /*return (
```
This is leftover, right?

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,383 @@
+  addMessages(messageList){
+    var messages = _.clone(this.state.messages);
+    var initialResponse = this.state.initialResponse;
+    var revisedResponse = this.state.revisedResponse;
```
Switch these `var`s over to `const`s where you can do indicate that they don't get mutated at all.  You can also use destructuring like this to simplify a bit:

```
const {initialResponse, revisedResponse} = this.state;
```

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,383 @@
+      if(message.text !== ''){
+        messages.push(message);
+      }
+      if(message.type === 'user'){
```
This looks like it takes a list of messages and then extracts whether there was an `initialResponse` or `revisedResponse` in there.  Could this be done at the call site?

In order words, in the method handling when the user types in the revised response message, would it work to have that `setState` to add a message and then also setState for the `revisedResponse`?

Alternately we could use only `messages` with the different types you have and write a method to pull out the last `revisedResponse` and remove that from state altogether.  Essentially, keep all that state in messages.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,383 @@
+    this.props.resetExperience();
+  },
+
+  getInitialMessages(nextProps){
```
I'm not quite following what this method is doing, is it checking the props and then figuring out what initial messages to show based on the question?

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,383 @@
+  },
+  
+  componentWillMount(){
+    this.messageId = parseInt(this.props.question.id)*10 + 1;
```
I'm not quite sure what this is about - do messages need an id in order to use it as a key when rendering?  I was thinking that the message index might be sufficient for that purpose.  Or, if you want to make a unique message id, you can used the `uuid` package.  I'm not sure why it's related to the `question.id` here though.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,383 @@
+    this.messageId = parseInt(this.props.question.id)*10 + 1;
+  },
+  
+  componentWillReceiveProps(nextProps){
```
In general, React components shouldn't use `componentWillReceiveProps` and `componentDidUpdate`.  These are primarily for interop with other libraries or dropping down to do raw DOM manipulations.  You shouldn't have to do that in most normal UI development.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,383 @@
+    const badExamples = _.map(nextProps.question.nonExamples, (example) => {
+      return {type: 'Bad', text: example};
+    });
+    const allExamples = _.shuffle(_.clone(goodExamples).concat(badExamples)); 
```
Why are the examples getting re-shuffled here?

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,383 @@
+      return {type: 'Bad', text: example};
+    });
+    const allExamples = _.shuffle(_.clone(goodExamples).concat(badExamples)); 
+    //this.addMessages(messages);
```
leftover comment

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,383 @@
+  
+  componentDidUpdate(prevProps, prevState){
+    if(prevState.elapsedMs === 90000){
+      this.addMessages([{type: 'info', text: 'Remember, these are supposed to be quick responses.'}]);
```
This seems more natural to put in `updateTimer`, rather than after every state update.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,383 @@
+      this.logRevisionDeclined();
+    }else if(prevState.revisedResponse === undefined && this.state.revisedResponse !== undefined){
+      this.logRevision(this.state.revisedResponse);
+    }
```
Let's see if we can do this at the call site, where we handle the action that the user is taking, instead of having to compute what has changed here.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,383 @@
+            <IconButton
+              onTouchTap={this.resetExperience}
+            >
+              {!hasStarted &&
```
This seems like it could be more readable as one line since these are so short and wouldn't line-wrap, up to you.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,383 @@
+          }
+          />
+        <div>
+        {/*Header*/}
```
leftover comment

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,383 @@
+          
+        </div>
+        <div>
+        {/*Scollable Body*/}
```
leftover comment

kevinrobinson commented on ui/src/message_popup/mobile_prototype/texting_interface.jsx:
```diff
@@ -0,0 +1,154 @@
+import {StudentMessage, UserMessage, InfoMessage} from './message_components.jsx';
+
+export const TextBody = React.createClass({
+  propTypes: {
```
Please add in a `displayName` property for all React components so they look nice in debugging messages.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/texting_interface.jsx:
```diff
@@ -0,0 +1,154 @@
+  },
+  
+  onPassSendMessage(){
+    this.props.addResponse('');
```
This seems like it's a different user action and important to capture - passing on revising vs. adding a blank response.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/texting_interface.jsx:
```diff
@@ -0,0 +1,154 @@
+  },
+  
+  onNextQuestion(){
+    this.setState(this.getInitialState());
```
In general this isn't a great pattern in React - it's better to update state and then let new components be created and destroyed based on what `render` returns rather than managing the state changes like this manually.  That's one of the main benefits, that you often don't have to think of resetting state.

Here, if we remove this what would happen?

kevinrobinson commented:
@david-rosales Nice work on taking an initial product concept, sketching it out and then prototyping parts of it!

Here's my suggestions for next steps, echoing some of what we talked about in person:
1. there's too much here to try to get this merged for the playtest tonight, which is fine
2. let's finish the refactoring work in https://github.com/mit-teaching-systems-lab/threeflows/pull/67 first
3. let's try to clarify the central texting metaphor in some design sketches next
4. then we can do an internal playtest on those sketches

In terms of the code here, I think it'd be good to sync on `componentDidUpdate` and `componentDidReceiveProps` and the flow of state through the components.  Let's do that in person whenever you think is a good time.

kevinrobinson commented:
@david-rosales Thanks for sharing this and talking it out in person, your screenshots look like they're ready for some playtesting!  👍 

You can put them in a Slides presentation and then show people in person one-on-one and get feedback.  Then you can take the input from a few people in person and synthesize it to make your plan for what the next steps should be.

david-rosales commented:
I thought that I'd document a few of the changes I've done to the Mobile Prototype since the initial build. One of them is that student information is now displayed by touching on the student's icon or through an information icon.

Touching on an information icon will bring up this Dialog:
<img width="270" alt="screen shot 2016-07-28 at 11 43 44 am" src="https://cloud.githubusercontent.com/assets/1490352/17219750/78d7d330-54ba-11e6-9d93-d44aae430787.png">
The reasoning behind using the information icon as well is because not all questions will have students speaking. Some will solely be situational while there still is a student involved. In those cases, accessing the student's information is still important.

From the Involved Students Dialog, you can tap on a student chip or from the texting interface, you can tap on a student's icon to display their information like below:
<img width="270" alt="screen shot 2016-07-28 at 11 44 19 am" src="https://cloud.githubusercontent.com/assets/1490352/17219826/d1dc21b6-54ba-11e6-882e-865ffcc04292.png">

Another addition is that there is now support for multiple students. I originally added this feature here but as it came up in a separate pull request I decided to redo it on its own branch so that it could be shared between in progress pull requests. The difference here being that there is `encodedText` which works as follows:
- If the text is informational, enter the following tag before it: `[:i]`.
- If the text is from the user, enter the following tag before it: `[:u]`.
- If the text is from a student, enter the following tag before it: `[:{studentId}s]`. For example, if Ada was the student, Ada's ID is 8 and so the tag would be `[:8s]`.
- The text after the tag will be in the type until the next tag.

With multiple students also comes the feature that students now speak what they "say" in the question. The `encodedText` is a property of a question just as `text` or `id` is. Furthermore, the `studentId` Number property has been changed to a `studentIds` Array property to support multiple students.

Below are two examples of `encodedText` in action:
<div style="display: flex; justify-content: space-between">
<img width="270" alt="screen shot 2016-07-28 at 11 48 40 am" style="float: left;" src="https://cloud.githubusercontent.com/assets/1490352/17220050/dac00008-54bb-11e6-9883-5dffa509781a.png">
<img width="270" alt="screen shot 2016-07-28 at 11 49 19 am" style="float: right;" src="https://cloud.githubusercontent.com/assets/1490352/17220054/ddb2b008-54bb-11e6-93fd-f98b63d48d10.png">
</div>
The `encodedText` for the first image is:
```
encodedText: '[:i]You introduced a game related to photosynthesis, and have pairs of two working on a game. Maia is paired with Kevin.[:3s]How are we getting graded on this?[:u]There will be no grade for either completion or participation.[:3s]Why are we doing this then?'
```
The `encodedText` for the second image is:
```
encodedText: '[:i]At the end of the photosynthesis unit, you introduce a list of topics that students choose for a project they want to do. Maia raises her hand.[:3s]I don\'t want to do any of these. Can I just come up with my own thing?[:i]Even before you have a chance to respond to her, another student raises her hand. [:8s]I want to do my own topic too!'
```

Another thing that was added because of the multiple students is names near the top of each message bubble. This way, if there are multiple students speaking, like in the second image above, even though the students have the same icon, they are still differentiable.
<img width="362" alt="screen shot 2016-07-28 at 11 50 18 am" src="https://cloud.githubusercontent.com/assets/1490352/17220447/3d4f0e2a-54bd-11e6-8886-1aaeedb0001f.png">


Furthermore, since there are multiple students, you can also see multiple chips in the Involved Students Dialog.
<div style="display: flex; justify-content:space-between; flex-flow: column || nowrap;">
<img width="270" alt="screen shot 2016-07-28 at 11 50 55 am" src="https://cloud.githubusercontent.com/assets/1490352/17220497/83ed4504-54bd-11e6-8db5-4f983c82d363.png">
<img width="270" alt="screen shot 2016-07-28 at 12 28 49 pm" src="https://cloud.githubusercontent.com/assets/1490352/17220818/f906396c-54be-11e6-89c0-37996968efda.png">
</div>

Another issue I ran into was how to communicate to users that tapping on icons did things. One suggestion was to state it somewhere on the instructions page.
<img width="270" alt="screen shot 2016-07-28 at 11 54 41 am" src="https://cloud.githubusercontent.com/assets/1490352/17220912/6c9adb30-54bf-11e6-8e3c-ffcae6199543.png">
 The only issue with this approach is that most people regardless of how much it is stressed to do so, don't read instructions. So in order to make it clearer that tapping on icons was a feature, I added in a Final Reminders Dialog that appears when users click start.
<img width="270" alt="screen shot 2016-07-28 at 11 55 19 am" src="https://cloud.githubusercontent.com/assets/1490352/17220993/c7578834-54bf-11e6-95d1-d5e907c67872.png">
This should be enough to communicate the necessary information to users. However, one other suggestion I received that I experimented with was making the icons look like they were clickable by giving them a subtle animation. I tried this but ran into the issue of trying to decide which icons to animate. Also, if I was animating every icon that was clickable, all the icons on the left would be animated and in unison. The movement in unison was extremely distracting and I think takes away from the overall experience so I did not pursue it. However, if we find that the above two attempts at attracting user attention are insufficient, we can try again with different animations.

On another note, animations have been added to the messages so that they appear to pop in when "sent". 

Lastly, I updated the feedback view to be solely information text (compared to the change between information and student text that was originally implemented). The feedback/revision message now looks as follows:
<img width="270" alt="screen shot 2016-07-28 at 11 53 58 am" src="https://cloud.githubusercontent.com/assets/1490352/17221170/a80b9708-54c0-11e6-9676-13fae6eb4edb.png">

One other thing we need to do which can be in a separate pull request is, for the new questions, assign students that are in the virtual classroom.

This is an update/overview of the changes since the first review. I might have missed a few small things and will add them if I recall/come across them.

kevinrobinson commented on ui/src/message_popup/experience_page.jsx:
```diff
@@ -127,9 +136,16 @@ export default React.createClass({
               primaryText="Restart game" />
           }
         />
+        <Dialog
```
We talked about removing this, but also as it is written now this would affect the simple "non-mobile prototype" interface too.  When making larger changes like this it's best if we can isolate the changes to only the new UI code, like within `renderMobilePrototype` or code it uses.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/message_components.jsx:
```diff
@@ -0,0 +1,178 @@
+import FaceIcon from 'material-ui/svg-icons/action/face';
+import InfoOutlineIcon from 'material-ui/svg-icons/action/info-outline';
+
+const Message = React.createClass({
```
This is a great use of the `Message` component as a small piece that other components use. 👍 

If you choose to have multiple components in a file like this, please add a quick comment at the top of the file about what this file contains and what's going on.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/message_components.jsx:
```diff
@@ -0,0 +1,178 @@
+  },
+  
+  componentDidMount(){ 
+    ReactDOM.findDOMNode(this).scrollIntoView(); 
```
Will this work as expected if multiple `Messages` are on the page?

kevinrobinson commented on ui/src/message_popup/mobile_prototype/message_components.jsx:
```diff
@@ -0,0 +1,178 @@
+    
+    return (
+      <div style={messageStyle}>
+        { type === 'student' && 
```
It seems like the core idea here is that `Message` can have a different icon, or maybe no icon.  Instead of putting the details of the `StudentMessage` and `InfoMessage` inside here, could you update `Message` to take an `icon` and then pass in those values from the particular message components?

kevinrobinson commented on ui/src/message_popup/mobile_prototype/message_components.jsx:
```diff
@@ -0,0 +1,178 @@
+    padding: 0,
+    flexGrow: 0,
+    flexShrink: 0,
+    flexBasis:30
```
Does it make sense for this and the other icons to have a fixed width and height?

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,375 @@
+  
+  getInitialState() {
+    var allExamples = [];
+    var messages = [];
```
Can `allExamples` and `messages` be constant?

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,375 @@
+    return ({
+      messages,
+      allExamples,
+      selectedStudent: this.props.question.students[0],
```
Is this just used for the dialog?  Should it be part of that, something like `setState({ dialog: { type: 'student', studentId: 5 })`?

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,375 @@
+      revisedResponse: undefined,
+      elapsedMs: 0,
+      exampleIndex: 0,
+      dialog: 'none',
```
Would `null` make sense here?

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,375 @@
+      elapsedMs: 0,
+      exampleIndex: 0,
+      dialog: 'none',
+      messageId: 1
```
This looks like an index into the `messages` array rather than a primary key that you'd use to look up a message, so i'd suggest `messageIndex` instead.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,375 @@
+    this.setState({messages, messageId});
+  },
+  
+  addResponse(response){
```
This can make sense as a method, but if it's for adding `user` messages then the name should reflect that and you'll want to make sure you call it each time.  As is, some methods use this to add a user message but others create them directly, so I'd either update to have them all use this or remove it.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,375 @@
+    }else{
+      this.addMessages([
+        {type: 'user', text: response},
+        {type: 'info', text: "There aren't any examples for this situation yet but your response will be recorded for feedback."}, 
```
I'd cut this line altogether, there may not actually be feedback for many of the testing sessions.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,375 @@
+    }
+  },
+  
+  showExample(){
```
As a convention, please update this method and others in this component to indicate they're a response to an event.  For this it might be `onShowExampleClicked`.  This can help clarify which methods are event handlers and which ones are side-effect free and just calculate something from state (eg., `isReadyToMoveOn`).

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,375 @@
+    }
+  },
+
+  setDialog(value){
```
I'd remove this, I'm not sure it's adding much to say `setDialog(value)` over `setState({dialog: value})`.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,375 @@
+    return (function() {
+      this.setState({selectedStudent: student});
+      this.setDialog('student');
+    }.bind(this));
```
Do you need this `bind` here?

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,375 @@
+    this.props.onLog(type, response);
+  },
+
+  getInitialMessages(questionObject, shouldShowStudentCard){
```
Does this use `shouldShowStudentCard`?  Looks like the linter settings don't pick this up and they should, I'll update that separately.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,375 @@
+  getInitialMessages(questionObject, shouldShowStudentCard){
+    var messages = [];
+    if(_.has(questionObject, 'encodedText')){
+      var arrayOfMessages = questionObject.encodedText.split('[:');
```
Encoding this structure in the questions directly will let you remove this parsing code altogether :)

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,375 @@
+                setInitialResponse={this.setInitialResponse}
+                setRevisedResponse={this.setRevisedResponse}
+                setPassedResponse={this.setPassedResponse}
+                log={{logResponse: this.logResponse, logRevision: this.logRevision, logRevisionDeclined: this.logRevisionDeclined}}
```
Nice wrapping this up into a logical object!

kevinrobinson commented on ui/src/message_popup/mobile_prototype/mobile_interface.jsx:
```diff
@@ -0,0 +1,375 @@
+                />
+            </div>
+            <div>
+              {selectedStudent !== undefined &&
```
This looks good, but I'd recommend splitting out the rendering of the dialogs to their own method.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/texting_interface.jsx:
```diff
@@ -0,0 +1,184 @@
+
+import {StudentMessage, UserMessage, InfoMessage} from './message_components.jsx';
+
+export const TextBody = React.createClass({
```
Would it work to split these two components into their own files?  If not please add a quick comment at the top of the file explaining what's in here.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/texting_interface.jsx:
```diff
@@ -0,0 +1,184 @@
+    messages: React.PropTypes.array.isRequired,
+    onOpenStudentDialog: React.PropTypes.func.isRequired,
+    onOpenInfoDialog: React.PropTypes.func.isRequired
+  },
```
Awesome API to the component, super clear!

kevinrobinson commented on ui/src/message_popup/mobile_prototype/texting_interface.jsx:
```diff
@@ -0,0 +1,184 @@
+    onOpenInfoDialog: React.PropTypes.func.isRequired
+  },
+  
+  render(){
```
Nice!  This is great factoring to pull out the components of the messages here, it makes this clear and direct.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/texting_interface.jsx:
```diff
@@ -0,0 +1,184 @@
+    const {messages} = this.props;
+    return (
+      <div>
+        {_.map(messages, (message) => { 
```
You can use `message.map` directly, Arrays have map built in.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/texting_interface.jsx:
```diff
@@ -0,0 +1,184 @@
+            rows={2}
+            rowsMax={2}
+            onChange={this.onTextChange}
+            disabled={this.props.isReadyToMoveOn()}
```
Instead of passing down the function `isReadyToMoveOn`, pass the value directly.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/texting_interface.jsx:
```diff
@@ -0,0 +1,184 @@
+    marginRight: 5
+  },
+  responseTimer: {
+    position: 'fixed',
```
Can you use flexbox here?  I don't think we want fixed position for anything if we can avoid it.

kevinrobinson commented:
@david-rosales Awesome work!  I left some comments, let me know if you want to chat about any of them or if it's helpful for me to do a final review on this pass.

kevinrobinson commented:
@david-rosales Also, here are the raw notes from our earlier in-person conversation, which we broke up into three pull requests:

```
prompt about tapping on students: secondary so remove
end button -> respond to Maia
...
encodedText - encode structure directly
...
animation to side not grow in, could use stagger
```

david-rosales commented on ui/src/message_popup/mobile_prototype/message_components.jsx:
```diff
@@ -147,7 +151,7 @@ const styles = {
+    flexBasis: 30
   },
   messageIcon: {
     width: 30,
```
I tried changing the fixed width and height by relying on setting just the flexBasis and using margins to position it but the IconButton wave effect looks really awkward and was off-centered.

david-rosales commented:
@kevinrobinson I went through the changes you suggested, removed the Final Reminders dialog, and changed the `SEND` button to be more action oriented. For now, there are a few questions that, although they should, do not have studentIds associated with them. With no studentIds, the `SEND` button now defaults to `RESPOND TO CLASS` instead of `RESPOND TO [StudentNameHere]`. Can you review the new changes one last time?

david-rosales commented on ui/src/message_popup/mobile_prototype/message_components.jsx:
```diff
@@ -0,0 +1,182 @@
+  },
+  messageIcon: {
+    width: 30,
+    height: 30,
```
I tried changing the fixed width and height by relying on setting just the flexBasis and using margins to position it but the IconButton wave effect looks really awkward and was off-centered.

kevinrobinson commented on ui/src/message_popup/mobile_prototype/text_footer_component.jsx:
```diff
@@ -0,0 +1,145 @@
+  },
+  
+  render(){
+    const mainButtonLabel = this.props.isReadyToMoveOn ? this.props.nextButtonLabel : this.props.helpType !== 'feedback' ? 'Respond to ' + (this.props.mainStudent !== undefined ? this.props.mainStudent.name : 'Class') : this.state.initialMessageSent ? 'Revise Response' : 'Save Response';
```
This is a long line :)  Can you break this up a bit? Triply nested ternaries are usually harder to read than they're worth and there's a bug with the initial button text in practice mode at the moment that's hard to spot.

I'd suggest splitting this out into a method, and then checking for conditions that you can return early from.  Something like:

```
renderButtonText(): {
  if (this.props.isReadyToMoveOn) return this.props.nextButtonLabel;
  if (this.state.initialMessageSent && this.props.helpType === 'feedback') return 'Revise response';
  ...
}
```

kevinrobinson commented:
@david-rosales This looks great!  There's a small bug in the branching for different button texts, and then a merge conflict to resolve but then let's ship this.  Awesome work, this seems like an appropriate approval gif:

![spiderman texting](http://67.media.tumblr.com/c70f156d9f029f68d3c380c2bffb63dd/tumblr_o463kuomXz1s02vreo1_500.gif)

david-rosales commented:
@kevinrobinson I broke up the rendering of the button label text as you suggested but am still unsure of what the bug was. Can you clarify just to make sure that it has been fixed? Thanks

kevinrobinson commented on ui/src/message_popup/mobile_prototype/text_footer_component.jsx:
```diff
@@ -0,0 +1,151 @@
+    if(this.props.isReadyToMoveOn) return this.props.nextButtonLabel;
+    if(this.props.helpType === 'feedback' && this.state.initialMessageSent) return 'Revise response';
+    return `Respond to ${this.props.mainStudent !== undefined ? this.props.mainStudent.name : 'Class'}`;
+  }
```
@david-rosales This looks good to me!  And a good place where setting up tests will be helpful :)

Formatting the code from https://github.com/mit-teaching-systems-lab/threeflows/pull/83/commits/b019b831572b99e47bdc5ec447dd47f7e88a1a9a, the bug came up when the user is first shown the question in solution mode, before they've typed anything:

```
const mainButtonLabel = this.props.isReadyToMoveOn
  ? this.props.nextButtonLabel
  : this.props.helpType !== 'feedback'
    ? 'Respond to ' + (this.props.mainStudent !== undefined ? this.props.mainStudent.name : 'Class')
    : this.state.initialMessageSent
      ? 'Revise Response'
      : 'Save Response';
```

looks good now!

david-rosales commented:
Alright awesome, merging now.
